### PR TITLE
add sink lifecyle procedures

### PR DESCRIPTION
**Pull request by conker84**
Fixes #320 in branch 4.0

Provide the same set of feature as we did in branch `3.5`

## Proposed Changes (Mandatory)

A brief list of proposed changes in order to fix the issue:
- added `streams.sink.stop`
- added `streams.sink.start`
- added `streams.sink.restart`
- added tests

### CODE REVIEW

**Comment by jexp** [comment]
Are these meant to be per database or global?

File: common/src/main/kotlin/streams/utils/Neo4jUtils.kt [hunk 1]
```diff
@@ -37,6 +37,21 @@ object Neo4jUtils {
...
     }
 
+    fun hasApoc(db: GraphDatabaseAPI): Boolean {
+        try {
+            return db.execute("RETURN apoc.version() AS version") {
+                it.columnAs<String>("version").next()
+                true
+            }
+        } catch (e: QueryExecutionException) {
+            if (e.statusCode.equals("Neo.ClientError.Statement.SyntaxError", ignoreCase = true)
```
Comment (jexp) [inline common/src/main/kotlin/streams/utils/Neo4jUtils.kt:47]: too specific
I would return false in any error case

File: consumer/src/main/kotlin/streams/kafka/KafkaEventSink.kt [hunk 2]
```diff
@@ -137,7 +144,7 @@ class KafkaEventSink(private val config: StreamsConfig,
                         }
                         TimeUnit.SECONDS.toMillis(1)
                     } else {
-                        val timeMillis = TimeUnit.MINUTES.toMillis(5)
```
Comment (jexp) [inline consumer/src/main/kotlin/streams/kafka/KafkaEventSink.kt:140]: 5 or 3 minutes both sound quite long, or?

Comment (jexp) [reply consumer/src/main/kotlin/streams/kafka/KafkaEventSink.kt:140]: is that configureable?

File: consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt [hunk 3]
```diff
@@ -1,21 +1,34 @@
...
 import streams.config.StreamsConfig
+import streams.events.StreamsPluginStatus
+import streams.extensions.toPointCase
+import streams.serialization.JSONUtils
+import streams.utils.Neo4jUtils
+import java.util.stream.Collectors
 import java.util.stream.Stream
 
 class StreamResult(@JvmField val event: Map<String, *>)
+class StreamsSinkDTO(@JvmField val name: String, @JvmField val value: Any?)
```
Comment (jexp) [inline consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt:22]: that's a really odd name as it leaks implementation details into the API

File: consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt [hunk 4]
```diff
@@ -35,6 +48,80 @@ class StreamsSinkProcedures {
...
+                    Stream.of(StreamsSinkDTO("exception", ExceptionUtils.getStackTrace(e))))
+        }
+    }
+
+    @Procedure("streams.sink.stop")
+    fun sinkStop(): Stream<StreamsSinkDTO> {
+        checkEnabled()
+        checkLeader()
+        try {
+            streamsEventSink?.stop()
```
Comment (jexp) [inline consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt:70]: what happens with the status that is managed by the kernel-extension-factory?

Will it try to stop / start the sink again if it's already started or stopped? Can we make sure that still works.

.e.g. someone calls sink.stop and then the database triggers a stop, will that be handled gracefully

or a restart of the db in the cluster

File: consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt [hunk 5]
```diff
@@ -35,6 +48,80 @@ class StreamsSinkProcedures {
...
+            return stopped.stream()
+        }
+        return sinkStart()
+    }
+
+    @Procedure("streams.sink.config")
+    fun sinkConfig(): Stream<StreamsSinkDTO> {
+        checkEnabled()
+        checkLeader()
+        val configMap = JSONUtils.asMap(streamsSinkConfiguration)
```
Comment (jexp) [inline consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt:93]: are there any passwords/credentials that we need to filter out/mask?

File: consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt [hunk 6]
```diff
@@ -35,6 +48,80 @@ class StreamsSinkProcedures {
...
+                .filterKeys { it != "topics" && it != "enabled" && it != "proceduresEnabled" && !it.startsWith("check") }
+                .mapKeys { it.key.toPointCase() }
+                .mapKeys {
+                    when (it.key) {
+                        "error.config" -> "streams.sink.errors"
```
Comment (jexp) [inline consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt:98]: looks as if this config listing / renaming (parse / output) should be in a central place and not hidden away in the procedure

so if something changes there we need to change it only in one place.

File: consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt [hunk 7]
```diff
@@ -35,6 +48,80 @@ class StreamsSinkProcedures {
...
+                .mapKeys { it.key.key }
+        val invalidTopics = mapOf("invalid_topics" to streamsSinkConfiguration.topics.invalid)
+        return (configMap + topicMap + invalidTopics)
+                .entries.stream()
+                .map { StreamsSinkDTO(it.key, it.value) }
+    }
+
+    @Procedure("streams.sink.status")
+    fun sinkStatus(): Stream<StreamsSinkDTO> {
+        checkEnabled()
```
Comment (jexp) [inline consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt:113]: do these functions have any return values?

or will they just error if something goes wrong?

File: consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt [hunk 8]
```diff
@@ -35,6 +48,80 @@ class StreamsSinkProcedures {
...
+        checkLeader()
+        val value = (streamsEventSink?.status() ?: StreamsPluginStatus.UNKNOWN).toString()
+        return Stream.of(StreamsSinkDTO("status", value))
+    }
+
+    private fun checkLeader() {
+        if (!Neo4jUtils.isWriteableInstance(db as GraphDatabaseAPI)) {
```
Comment (jexp) [inline consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt:120]: should that be returned as an exception or as part of the status "error" -> "You can use it only in the LEADER or in a single instance configuration."

also what is `it`.
and this is another piece of code that should probably live in a more central place?

File: doc/asciidoc/procedures/index.adoc [hunk 9]
```diff
@@ -241,3 +241,64 @@ Supported deserializers are: `org.apache.kafka.common.serialization.ByteArrayDes
 
 |===
 
+=== Streams Sink Lifecycle procedure
```
Comment (jexp) [inline doc/asciidoc/procedures/index.adoc:244]: can we add a usage example of an (textual) output of such a lifecycle?

```
listconfig
stop
status
start
status
restart
status on non leader
```

Comment (jexp) [reply doc/asciidoc/procedures/index.adoc:244]: we can also generate that from a test an just adoc-include the file.

**Review by jexp — APPROVED** [review]
please have a look at my comments, looks good otherwise.

File: common/src/main/kotlin/streams/utils/Neo4jUtils.kt (hunk 1 above)
Comment (conker84) [reply common/src/main/kotlin/streams/utils/Neo4jUtils.kt]: done

File: consumer/src/main/kotlin/streams/kafka/KafkaEventSink.kt (hunk 2 above)
Comment (conker84) [reply consumer/src/main/kotlin/streams/kafka/KafkaEventSink.kt:140]: I'll add the configuration in the next PR where I'll add all the changes made in #344

File: consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt (hunk 3 above)
Comment (conker84) [reply consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt]: changed

File: consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt (hunk 4 above)
Comment (conker84) [reply consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt]: Great question! I'll manage this in the next PR where I'll add all the changes made in #344

File: consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt (hunk 5 above)
Comment (conker84) [reply consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt]: no, here we have only properties regarding `streams.*`

File: consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt (hunk 6 above)
Comment (conker84) [reply consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt]: moved into a method inside the `StreamsSinkConfiguration` class

File: consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt (hunk 7 above)
Comment (conker84) [reply consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt:105]: they just throw an exception if something goes wrong

File: consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt (hunk 8 above)
Comment (conker84) [reply consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt]: I changed it as you requested, but I left it here as is something specific for procedures

File: doc/asciidoc/procedures/index.adoc (hunk 9 above)
Comment (conker84) [reply doc/asciidoc/procedures/index.adoc:244]: done

**Comment by conker84** [comment]
> Are these meant to be per database or global?

per database

**Comment by conker84** [comment]
@jexp I updated the PR with your feedback. Thank you!

**Comment by moxious** [comment]
OK to merge?

File: consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt [hunk 10]
```diff
@@ -35,6 +50,71 @@ class StreamsSinkProcedures {
...
+                log?.error("Cannot stopped the Sink because of the following exception", e)
+                Stream.concat(sinkStatus(),
+                        Stream.of(KeyValueResult("exception", ExceptionUtils.getStackTrace(e))))
+            }
+        }
+    }
+
+    @Procedure("streams.sink.restart")
+    fun sinkRestart(): Stream<KeyValueResult> {
+        val stopped = sinkStop().collect(Collectors.toList())
```
Comment (jexp) [inline consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt:85]: this could be simplified but is not important

**Review by jexp — APPROVED** [review]

**Pull request merged**